soc: renesas:ra Support minimal renesas RA8 MCU PM

duynguyenxa opened the pull request:
This PR is aim to create a support for basic of PM on RA8 MCUs
+ This PR add the support of system timer by using ULPT (ultra-low power timer) in PM mode
+ Support power.c with 2 PM state.

thaoluonguw commented on drivers/timer/ra8_ulpt_timer.c:
```diff
@@ -0,0 +1,188 @@
+	/* Initializes the module. */
+	err = R_ULPT_Open(&g_ulpt_ctrl, &g_ulpt_cfg);
+	if (err != 0) {
+		return err;
```
@duynguyenxa : Should we have an __ASSERT to check FSP error and inform the error log?

duynguyenxa replied:
@thaoluonguw , it's fixed please recheck

thaoluonguw commented on modules/Kconfig.renesas_fsp:
```diff
@@ -17,3 +17,20 @@ config USE_RA_FSP_DTC
 	help
 	  Enable RA FSP DTC driver
+
+config USE_RA_FSP_SCE
+	bool
+	help
+	  Enable RA FSP SCE driver
```
@duynguyenxa : Is it redundant?

duynguyenxa replied:
@thaoluonguw , it's fixed

nordicjm commented on modules/Kconfig.renesas_fsp:
```diff
@@ -48,3 +48,20 @@ config HAS_RENESAS_RA_RSIP_E51A
 
 endif
+
+config USE_RA_FSP_SCE
+	bool
+	help
+	  Enable RA FSP SCE driver
+
+config USE_RA_FSP_TIMER_ULPT
+	bool
+	help
+	  Enable RA FSP TIMER-ULPT driver
+
+config USE_RA_FSP_LPM
+	bool
+	default y
+	depends on PM
+	help
+	  Enable RA FSP LPM driver
```
why does nothing in this file have an if or depends on?

duynguyenxa replied:
Thank you for checking, it's fix in https://github.com/zephyrproject-rtos/zephyr/pull/78208

nordicjm commented on boards/renesas/ek_ra8d1/Kconfig.defconfig:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 32768 if ULPT_RENESAS_RA
+	default 480000000
```
why is this in board config? This is a SoC config item, it should be in soc/*/Kconfig.defconfig

duynguyenxa replied:
@nordicjm . I moved it in the zephyr/soc/renesas/ra/ra8*/Kconfig.defconfig

nordicjm commented on drivers/timer/ra8_ulpt_timer.c:
```diff
@@ -0,0 +1,188 @@
+#include <zephyr/spinlock.h>
+#include <zephyr/sys_clock.h>
+
+#define SYS_CLOCK_HW_CYCLES_PER_SEC (32768U)
```
you can't just assume this, get it from Kconfig and if it needs a specific value then #error out

duynguyenxa replied:
@nordicjm , This is redundant and not used anywhere, I remove it, the actual frequency for calculation is get from sys_clock_hw_cycles_per_sec()

nordicjm commented on soc/renesas/ra/ra8d1/power.c:
```diff
@@ -0,0 +1,135 @@
+lpm_instance_ctrl_t g_lpm_sw_standby_ctrl;
+const lpm_cfg_t g_lpm_sw_standby_cfg = {
+	.low_power_mode = LPM_MODE_STANDBY,
+	.standby_wake_sources = LPM_STANDBY_WAKE_SOURCE_IRQ13 | LPM_STANDBY_WAKE_SOURCE_ULP0U |
+				(lpm_standby_wake_source_t)0,
```
why is there a definitive IRQ13 here? Why does this not come from DTS? Having had a quick look, I think this is a UART interrupt, so I'm even more confused why this is here?

duynguyenxa replied:
@nordicjm , Thank for pointing out, the source code was not clean fully, at the beginning the IRQ13 aim to be external interrupt for testing. I removed it.

nordicjm replied:
It's still here?

thaoluonguw commented on drivers/timer/CMakeLists.txt:
```diff
@@ -28,6 +28,7 @@ zephyr_library_sources_ifdef(CONFIG_NPCX_ITIM_TIMER npcx_itim_timer.c)
 zephyr_library_sources_ifdef(CONFIG_NRF_GRTC_TIMER nrf_grtc_timer.c)
 zephyr_library_sources_ifdef(CONFIG_NRF_RTC_TIMER nrf_rtc_timer.c)
 zephyr_library_sources_ifdef(CONFIG_RCAR_CMT_TIMER rcar_cmt_timer.c)
+zephyr_library_sources_ifdef(CONFIG_ULPT_RENESAS_RA ra8_ulpt_timer.c)
```
Please change file name to "renesas_ra8_ulpt_timer" to easy detect for Renesas RA maintaining.

duynguyenxa replied:
@thaoluonguw , It's fixed

thaoluonguw commented on drivers/timer/Kconfig:
```diff
@@ -97,6 +97,7 @@ source "drivers/timer/Kconfig.xlnx_psttc"
 source "drivers/timer/Kconfig.xtensa"
 source "drivers/timer/Kconfig.mtk_adsp"
 source "drivers/timer/Kconfig.sy1xx_sys_timer"
+source "drivers/timer/Kconfig.ra8_ulpt"
```
Please change file name to easy detect for Renesas RA maintaining.

duynguyenxa replied:
@thaoluonguw , It's fixed

thaoluonguw commented on drivers/timer/ra8_ulpt_timer.c:
```diff
@@ -0,0 +1,183 @@
+	err = R_ULPT_Open(&g_ulpt_ctrl, &g_ulpt_cfg);
+	__ASSERT(err == 0, "ULPT timer: initialization: open failed");
+
+	R_ULPT_Start(&g_ulpt_ctrl);
```
Lack of "err"

duynguyenxa replied:
@thaoluonguw , It's fixed

KhiemNguyenT reviewed (dismissed):
All the comments have been resolved, so I approve.

nordicjm commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -22,12 +22,26 @@
+
+		power-states {
+			stop0: state0 {
+					compatible = "zephyr,power-state";
+					power-state-name = "runtime-idle";
+					min-residency-us = <300>;
+				};
+			stop1: state1 {
+					compatible = "zephyr,power-state";
+					power-state-name = "standby";
+					min-residency-us = <50000>;
+				};
```
alignment wrong on end brackets, indent wrong on middle lines

duynguyenxa replied:
It's fixed

nordicjm commented on soc/renesas/ra/ra8d1/Kconfig.defconfig:
```diff
@@ -13,4 +13,14 @@ config PINCTRL
+config SYS_CLOCK_TICKS_PER_SEC
+	default 4096 if ULPT_RENESAS_RA
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 32768 if ULPT_RENESAS_RA
+	default 480000000
```
@gmarull ?

nordicjm replied:
this needs to come from a dts property

nordicjm commented on soc/renesas/ra/ra8t1/power.c:
```diff
@@ -0,0 +1,135 @@
+lpm_instance_ctrl_t g_lpm_sw_standby_ctrl;
+const lpm_cfg_t g_lpm_sw_standby_cfg = {
+	.low_power_mode = LPM_MODE_STANDBY,
+	.standby_wake_sources = LPM_STANDBY_WAKE_SOURCE_IRQ13 | LPM_STANDBY_WAKE_SOURCE_ULP0U |
```
and here

duynguyenxa replied:
@nordicjm , Oops sorry I missed them, it's fixed. Thank you

ydamigos commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -198,6 +198,9 @@
 	interrupts = <40 1>, <41 1>;
 	interrupt-names = "gtioca", "overflow";
 	pinctrl-names = "default";
+};
```
pwm7 won't be enabled with this change. Is it intentional?

ydamigos commented on boards/renesas/mck_ra8t1/mck_ra8t1.dts:
```diff
@@ -138,6 +138,9 @@
 	interrupts = <40 1>, <41 1>;
 	interrupt-names = "gtioca", "overflow";
 	pinctrl-names = "default";
+};
```
pwm2 won't be enabled with this change. Is it intentional?

ydamigos commented on boards/renesas/ek_ra8m1/Kconfig.defconfig:
```diff
@@ -10,4 +10,8 @@ config NET_L2_ETHERNET
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 32768 if ULPT_RENESAS_RA
+	default 480000000
+
```
It was added at SoC level, do we still need this?

duynguyenxa replied:
It's removed in latest update

ydamigos commented on soc/renesas/ra/ra8t1/Kconfig.defconfig:
```diff
@@ -10,4 +10,14 @@ config NUM_IRQS
+	default 4096 if ULPT_RENESAS_RA
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 32768 if ULPT_RENESAS_RA
```
When ULPT is used, does M85 run on 32KHz?

duynguyenxa replied:
hi @ydamigos , sorry for taking too long to answer, no CM85 still running on CPU clock default config is 480MHz

ydamigos replied:
SYS_CLOCK_HW_CYCLES_PER_SEC option specifies the frequency of the hardware timer used for the system clock (in Hz).
If CM85 is not running at 32768 Hz we should not set it at this value even if `ULPT_RENESAS_RA` is set.

ydamigos replied:
So, CM85 will run at 32768 Hz, is that correct?

ydamigos replied:
@duynguyenxa ok clear. Thanks for the clarification.

khanhnguyenwz committed:
drivers: timer: Add ULPT timer for power management on Renesas RA MCUs

drivers:
- Added ULPT timer driver in `renesas_ra_ulpt_timer.c`.
- Updated `clock_control_renesas_ra_cgc.c` for ULPT clock settings.
- Updated `uart_renesas_ra8_sci_b.c` for power management support.
- Updated `CMakeLists.txt` and `Kconfig` to integrate ULPT timer.
- Added `Kconfig.renesas_ra_ulpt` for ULPT-specific configurations.

dts bindings:
- Added `renesas,ra-ulpt.yaml` for ULPT node bindings.
- Added `renesas,ra-ulpt-timer.yaml` for ULPT timer bindings.

modules:
- Updated `Kconfig.renesas_fsp` to support ULPT and LPM.

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
dts: arm: renesas: Add ULPT and power state nodes for Renesas RA8

ULPT nodes:
- Added `ulpt0` at 0x40220000 (channel 0).
- Added `ulpt1` at 0x40220100 (channel 1).
- Included timer sub-nodes with `renesas,ra-ulpt-timer` compatibility.

Power state nodes:
- `stop0`: Runtime idle state mapped to LPM sleep mode.
- `stop1`: Standby state mapped to LPM software standby mode.

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
soc: renesas: Add power management support for Renesas RA8

Updated `CMakeLists.txt` and `Kconfig` to integrate power management
for RA8D1, RA8M1, and RA8T1.

Modified `Kconfig.defconfig` to configure ULPT timer as the system timer
when power management is enabled:
- Adjusted `SYS_CLOCK_HW_CYCLES_PER_SEC` and `SYS_CLOCK_TICKS_PER_SEC`
for ULPT timer.
- Disabled `CORTEX_M_SYSTICK` when ULPT timer is used as the system timer.

Implemented power management logic in the new `power.c` file for:
- RA8D1 (`soc/renesas/ra/ra8d1/power.c`)
- RA8M1 (`soc/renesas/ra/ra8m1/power.c`)
- RA8T1 (`soc/renesas/ra/ra8t1/power.c`)

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
boards: renesas: Add power management support for RA8 boards

Updated DTS files to enable ULPT nodes for:
- `ek_ra8d1`
- `ek_ra8m1`
- `mck_ra8t1`

Updated board documentation in `index.rst` for each board.

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

nashif commented on tests/kernel/tickless/tickless_concept/boards/mck_ra8t1.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) 2025 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_USE_RENESAS_RA_ULPT_TIMER=y
```
why is this enabled only in the test and not set by the platform?

duynguyenxa replied:
@nashif , the configuration was unnecessary so I removed it. Thank you

nordicjm commented on drivers/timer/Kconfig.renesas_ra_ulpt:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2025 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config USE_RENESAS_RA_ULPT_TIMER
+	bool "Use the Renesas RA Series ULPT Timer"
```
lacks a depends on

duynguyenxa replied:
@nordicjm ,This config was unnecessary, so I removed it.

nordicjm commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -313,3 +313,17 @@ pmod_sd_shield: &sdhc1 {};
+
+&ulpt0 {
+	status = "okay";
+	timer {
```
newline spacing for these

duynguyenxa commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -22,12 +22,27 @@
+				power-state-name = "runtime-idle";
+				min-residency-us = <450>;
+			};
+			stop1: state1 {
```
@nordicjm , It's fixed, please take a look

thaoluonguw commented:
Please rebase to solve conflict.

duynguyenxa commented:
@nordicjm can you help to take a look again
